File: 0.1.1/loss_function/quadratic_loss.py
import numpy as np

def Function(output, sample):
    return (output-sample)**2

def Propagation_Unit(output, sample):
    return 2*(output-sample)

class Quadratic_Loss:

    def function(Q, start_position, batch_size, output_matrix, sample_matrix):
        loss_matrix = np.zeros(np.shape(output_matrix))
        for i in range(start_position, start_position+batch_size):
            for j in range(np.shape(loss_matrix)[1]):
                loss_matrix[i][j] = Function(output_matrix[i][j], sample_matrix[i][j])
        Q.put(np.sum(loss_matrix))

    def propagation(Q, start_position, batch_size, output_matrix, sample_matrix):
        propagation_units = np.zeros(np.shape(output_matrix))
        for i in range(start_position, start_position+batch_size):
            for j in range(np.shape(propagation_units)[1]):
                propagation_units[i][j] = Propagation_Unit(output_matrix[i][j], sample_matrix[i][j])
        Q.put(propagation_units.T)

File: 0.2.0/loss_function/quadratic_loss.jl
using .Threads

module Quadratic_Loss
    function opt_func(output::Float32, sample::Float32)
        return (output-sample)^2
    end

    function opt_pu(output::Float32, sample::Float32)
        return 2*(output-sample)
    end

    function func(output_matrix::Array{Float32}, sample_matrix::Array{Float32})
        loss_matrix = zeros(Float32, size(output_matrix))
        Threads.@threads for i in eachindex(loss_matrix)
            loss_matrix[i] = opt_func(output_matrix[i], sample_matrix[i])
        end
        return loss_matrix
    end

    function prop(output_matrix::Array{Float32}, sample_matrix::Array{Float32})
        propagation_units = zeros(Float32, size(output_matrix))
        Threads.@threads for i in eachindex(propagation_units)
            propagation_units[i] = opt_pu(output_matrix[i], sample_matrix[i])
        end
        return propagation_units
    end
end
